Coretime Guides Rework (Omninode Considerations) (#5946)
w3f/polkadot-wiki: docs/build/build-guides-template-basic.md

:::warning Not a production ready guide.

> This guide is considered a moving document and currently uses the **Rococo** testnet. The Kusama
> relay chain can also be used in place of Rococo, as coretime is also enabled there. Polkadot will

Reviewer: ```suggestion
This guide is considered a moving document and currently uses the **Rococo** testnet. This guide is also applicable to the parachains on the Kusama
relay chain, as coretime is also enabled there. Polkadot will
```

enable agile coretime after it has been thoroughly tested on Kusama.

> **This guide assumes Polkadot SDK tag `polkadot-v1.15.1`**

Reviewer: ```suggestion
**This instructions on this guide are applicable for the Polkadot SDK repository with tag `polkadot-v1.15.1`**
```

registered accounts and associated balances within the Polkadot.js Extension. Once finished, you
should see your new ParaId at the bottom of the list within
[Network > Parachains > Parathreads](https://polkadot.js.org/apps/#/parachains/parathreads) with the
option to "Deregister" to the right:

![Registered ParaID in PolkadotJs](../assets/coretime/Coretime-ParaId-Registered.png)

## Compiling Parachain Runtime and Generating Wasm Blob

We can now move on to working with the template. Essential prerequisites are:

1. **Have** a command line, git, and other common development tools to edit code/files.
2. **Rust**, its associated tooling, the nightly toolchain, and the `wasm32-unknown-unknown`
   compilation target.

:::info Install dependencies

Visit [the dependencies' installation](./build-guides-install-deps.md) page before starting.

:::

This guide uses release
[`polkadot-v1.15.0`](https://github.com/paritytech/polkadot-sdk/releases/tag/polkadot-v1.15.0), for
associated tooling (such as `polkadot-parachain` and `chain-spec-builder`).

We will be using the
[Polkadot SDK's parachain template](https://github.com/paritytech/polkadot-sdk-parachain-template),

> which is kept in sync with the Polkadot SDK.

Reviewer: ```suggestion
which is mirrored in the templates folder within Polkadot SDK repository.
```

FRAME and Substrate. All you need to know is the following:

- `runtime/` - Contains the runtime and business logic. This is how all of your pallets (runtime
  modules) are configured. The runtime, once it's compiled as a WebAssembly blob, is uploaded
  on-chain.
- `node/` - The node implementation, which takes care of networking and RPC setup.

> Pallets are essentially just Rust crates, which are imported as dependencies, as seen in
> `runtime/Cargo.toml`. Read more about
> [pallets here.](https://paritytech.github.io/polkadot-sdk/master/polkadot_sdk_docs/polkadot_sdk/frame_runtime/index.html#pallets)

When we compile our template, we can extract the runtime code as a WebAssembly `.wasm` blob, which
is one of the key artifacts for registering our parachain on the relay chain.

Build the node using the following command:

```shell
cargo build --release
```

For the sake of this example, we won't go into adding or modifying any pallets. However, this is
definitely a next step after you get used to deploying your parachain on Rococo!

### Customizing our chain specification's patch file

> The chain specification is a JSON file which describes Substrate-based networks. It usually contains

Reviewer: ```suggestion
The chain specification is a JSON file that describes Polkadot SDK-based networks. It usually contains
```

> the genesis runtime (in hex) under `genesis.runtimeGenesis.code`. Under that, it also contains
> genesis values/state for the pallets included in your runtime.

Reviewer: ```suggestion
the genesis runtime (in hex) under `genesis.runtimeGenesis.code` and also contains
genesis values/state for the pallets included in your runtime.
```

You can bootstrap your network with some initial values, such as initial collators, balances for
certain accounts, and more. This is done using a patch file, which the `chain-spec-builder` tool
uses to create the full chain specification of your network. You should do the following to create
your `patch.json`:

> Feel free to use the patch provided here, which you can look to tweaking to your liking.

Reviewer: ```suggestion
Feel free to use the patch provided here, which you can look to tweak to your liking.
```

    "tokenDecimals": 12,
    "tokenSymbol": "UNIT"
  },
  "para_id": YOUR_PARA_ID_HERE,
  "relay_chain": "rococo",
  "codeSubstitutes": {},
  "genesis": { ... }
}
```

Feel free to customize various aspects of your spec, such as the `UNIT` ticker, `name`, `id`, or
other fields.

Now you should open your `chain_spec.json`, and use this checklist to ensure all the necessary
fields are in place:

1. **Make** sure that `relay_chain` is set to the target relay chain (`rococo`, in our case)
2. **Make** sure that `para_id` (right below `relay_chain`) is set to your reserved ParaId
3. **Make** sure that our `chain_type` is set to `Live`
4. **Optionally**, change the name, id, and token symbol of your chain.

If you fail to do one of these, there is a large chance that your chain may fail to produce blocks.

For more information on chain specifications,
[check out the reference document from the Polkadot SDK.](https://paritytech.github.io/polkadot-sdk/master/polkadot_sdk_docs/reference_docs/chain_spec_genesis/index.html)

### Generating the Runtime and Genesis

With our chain specification successfully generated, we can move to generating the genesis state and
runtime.

> Generate the genesis as follows:

Reviewer: ```suggestion
Generate the genesis following the instructions below:
```

```shell
polkadot-parachain export-genesis-head --chain chain_spec.json genesis
```

Although you can use the WebAssembly within `wbuild`, for ease of access you can also regenerate
your WebAssembly blob with the following:

```shell
polkadot-parachain export-genesis-wasm --chain chain_spec.json genesis-wasm
```

Within your project folder, you should now have two files:

- **`genesis`** - the initial state of your parachain.
- **`genesis-wasm`** - the initial runtime WebAssembly blob of your parachain.

## Running Your Collator

> Make sure you have the
> [`polkadot-parachain`](./build-guides-install-deps.md#installing-the-omninode) binary installed!

> You should start syncing your collator. Keep in mind that you will need to sync Rococo first - this

Reviewer: This paragraph could improve with a bit of elaboration on the sync process. What do `blocks-pruning` and `state-pruning` do? even a small context on these flags will improve the reader's experience.

The tone is conversational and could be a bit more instructional. It's better to avoid terms like ASAP in technical documentation.

"could take some time depending on your download speed" -> "the sync process can benefit from higher network bandwidth as it is network resource intensive" or something along those lines

your WebAssembly blob with the following:

```shell
polkadot-parachain export-genesis-wasm --chain chain_spec.json genesis-wasm
```

Within your project folder, you should now have two files:

- **`genesis`** - the initial state of your parachain.
- **`genesis-wasm`** - the initial runtime WebAssembly blob of your parachain.

## Running Your Collator

> Make sure you have the
> [`polkadot-parachain`](./build-guides-install-deps.md#installing-the-omninode) binary installed!

Before you are able to connect your collator, you must sync the relay chain. Depending on your
download speed, the time to sync may vary. In order to avoid storing the full state of the relay
chain, be sure to run with the appropriate pruning flags (`blocks-pruning` and `state-pruning`):

:::info Explaining `blocks-pruning` and `state-pruning`

A Polkadot SDK-based node has two pruning modes:

- `blocks-pruning` - Prunes block bodies (the list of extrinsics in the block) from a specified
  height (default: `256`)
- `state-pruning` - Prunes the overall state from a specified height

Both of these flags aid in reducing the amount of disk space taken up by the relay chain. Note that

> `state-pruning` is only be used for the first initial sync for the database.

Reviewer: ```suggestion
`state-pruning` is only used for the first initial sync for the database.
```

deployment. This function should replace the `local_testnet_config` function within
`node/src/chain_spec.rs`:

> > **:warning: WARNING! :warning:** Make sure you replace the ParaId with the one you reserved

Reviewer 2: Yesssss, happy to see all of this goes away now 😇

Author: There is still some cleanup to do on my end, I mostly need to back around to the PR to allow for an easier way to generate parachain specs - although, it's just really modifying a JSON after generating it ATM, which isn't terrible, so I may push it with that extra manual step.  WDYT?

Reviewer 2: agree, I think editing a JSON file is well within the capabilities of anyone. It is not amazing devEx tho, but we can improve that next.
